### PR TITLE
expression, executor: length(binary(m)) should return 'm' instead of the concrete string length(#3644)

### LINKED ISSUE #3644: `length` result is wrong after modify column length
## Description
Please answer these questions before submitting your issue. Thanks!

1. What did you do?
```
create table t(a int, b binary(22));
insert into t set b='test';
alter table t modify b binary(33);
insert into t set b='test1';
select length(b) from t;
```

2. What did you expect to see?
```
+-----------+
| length(b) |
+-----------+
|        33 |
|        33 |
+-----------+
```

3. What did you see instead?
```
+-----------+
| length(b) |
+-----------+
|        22 |
|        33 |
+-----------+
```

4. What version of TiDB are you using (`tidb-server -V`)?

## SIG slack channel
[#sig-exec](https://tidbcommunity.slack.com/archives/CMRD79DRR)

## Score
- 300

## Mentor
 * @qw4990

### PR DESCRIPTION
**Pull request by zz-jason**

### CODE REVIEW

File: expression/builtin_string.go [hunk 1]
```diff
@@ -132,19 +133,33 @@ type lengthFunctionClass struct {
...
 	bf, err := newBaseBuiltinFuncWithTp(args, ctx, tpInt, tpString)
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
 	bf.tp.Flen = 10
-	sig := &builtinLengthSig{baseIntBuiltinFunc{bf}}
-	return sig.setSelf(sig), errors.Trace(c.verifyArgs(args))
+	var sig builtinFunc
+	argTp := args[0].GetType()
+	if argTp.Tp == mysql.TypeString && mysql.HasBinaryFlag(argTp.Flag) {
```
Comment (XuHuaiyu) [inline expression/builtin_string.go]: need varbinary be treated as this?

Comment (zz-jason) [inline expression/builtin_string.go]: no, just binary

```sql
DROP TABLE IF EXISTS table_binary;
CREATE TABLE table_binary (
    a BINARY(20),
    b VARBINARY(20),
    c TINYBLOB,
    d BLOB(20),
    e MEDIUMBLOB,
    f LONGBLOB,
    g BIGINT,
    UNIQUE KEY(a),
    UNIQUE KEY(g)
);
INSERT INTO table_binary (a, b, c, d, e, f, g) VALUES ("a ", "b ", "c ", "d ", "e ", "f ", 2);
```

```
select length(a), length(b), length(c), length(d), length(e), length(f), length(g) from table_binary;
+-----------+-----------+-----------+-----------+-----------+-----------+-----------+
| length(a) | length(b) | length(c) | length(d) | length(e) | length(f) | length(g) |
+-----------+-----------+-----------+-----------+-----------+-----------+-----------+
|        20 |         2 |         2 |         2 |         2 |         2 |         1 |
+-----------+-----------+-----------+-----------+-----------+-----------+-----------+
```

**Comment by hanfei1991** [comment]
LGTM

**Comment by XuHuaiyu** [comment]
This PR may not really fix #3644 .
In MySQL:
``` sql
create table t (a binary(10));
select length(concat(a, "123")) from t;
--> 13

alter table t modify a binary(20);
select length(concat(a, "123")) from t;
--> 23
```

In TiDB, we may still get 13 even after merging this PR.

**Pull request closed without merging**